lib/plugin/opsgenie/client.go

```diff
@@ -0,0 +1,267 @@
+		},
+	})
+	if conf.APIEndpoint != "" {
+		client.SetHostURL(conf.APIEndpoint)
+	} else {
+		client.SetHostURL("") // TODO: Find reasonable default
+	}
+	client.SetHeader("Authorization", "GenieKey "+conf.APIKey)
+	return OpsgenieClient{
+		client:           client,
+		clusterName:      clusterName,
+		webProxyURL:      webProxyURL,
+		defaultSchedules: conf.defaultSchedules,
+		Priority:         conf.Priority,
+	}, nil
+}
+
+// CreateAlert creates an opsgenie alert.
+func (og OpsgenieClient) CreateAlert(ctx context.Context, reqID string, reqData RequestData) (OpsgenieData, error) {
+	bodyDetails, err := og.buildAlertBody(reqID, reqData)
+	if err != nil {
+		return OpsgenieData{}, trace.Wrap(err)
+	}
+
+	body := AlertBody{
+		Message:     fmt.Sprintf("Access request from %s", reqData.User),
+		Alias:       fmt.Sprintf("%s/%s", ogAlertKeyPrefix, reqID),
+		Description: bodyDetails,
+		Responders:  og.getResponders(reqData),
+		Priority:    og.Priority,
+	}
+
+	var result AlertResult
+	if _, err = og.client.NewRequest().
+		SetContext(ctx).
+		SetBody(body).
+		SetResult(&result).
+		Post("alerts"); err != nil {
+		return OpsgenieData{}, trace.Wrap(err)
+	}
+
+	return OpsgenieData{
+		AlertID: result.Alert.ID,
+	}, nil
+}
+
+func (og OpsgenieClient) getResponders(reqData RequestData) []Responder {
+	schedules := og.defaultSchedules
+	if reqSchedules, ok := reqData.RequestAnnotations[ReqAnnotationRespondersKey]; ok {
+		schedules = reqSchedules
+	}
+	responders := []Responder{}
+	for _, s := range schedules {
+		responders = append(responders, Responder{
+			Type: "schedule",
+			ID:   s,
+		})
+	}
+	return responders
+}
+
+// PostReviewNote posts a note once a new request review appears.
+func (og OpsgenieClient) PostReviewNote(ctx context.Context, alertID string, review types.AccessReview) error {
+	note, err := og.buildReviewNoteBody(review)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	body := AlertNote{
+		Note: note,
+	}
+	if _, err := og.client.NewRequest().
+		SetContext(ctx).
+		SetBody(body).
+		SetPathParams(map[string]string{"alertID": alertID}).
+		SetQueryParams(map[string]string{"identifierType": "id"}).
+		Post("alerts/{alertID}/notes"); err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+// ResolveAlert resolves an alert and posts a note with resolution details.
+func (og OpsgenieClient) ResolveAlert(ctx context.Context, alertID string, resolution Resolution) error {
+	note, err := og.buildResolutionNoteBody(resolution)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	body := AlertNote{
+		Note: note,
+	}
+	if _, err := og.client.NewRequest().
+		SetContext(ctx).
+		SetBody(body).
+		SetPathParams(map[string]string{"alertID": alertID}).
+		SetQueryParams(map[string]string{"identifierType": "id"}).
+		Post("alerts/{alertID}/close"); err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+func (og OpsgenieClient) buildAlertBody(reqID string, reqData RequestData) (string, error) {
```

Comment:
These methods (`buildAlertBody`, `buildReviewNoteBody`, `buildResolutionNoteBody`) don't need to be a part of the `OpsgenieClient`. For `buildAlertBody` you can just pass proxy URL to it.